### PR TITLE
Various changes to existing diagnostics

### PR DESCRIPTION
**Pull request by estebank**
* [Add code to `invalid ABI` error, add span label, move list to help to make message shorter](https://github.com/rust-lang/rust/pull/51463/commits/23ae5af274defa9ff884f593e44a2bbcaf814a02):
```
error[E0697]: invalid ABI: found `路濫狼á́́`
  --> $DIR/unicode.rs:11:8
   |
LL | extern "路濫狼á́́" fn foo() {} //~ ERROR invalid ABI
   |        ^^^^^^^^^ invalid ABI
   |
   = help: valid ABIs: cdecl, stdcall, fastcall, vectorcall, thiscall, aapcs, win64, sysv64, ptx-kernel, msp430-interrupt, x86-interrupt, Rust, C, system, rust-intrinsic, rust-call, platform-intrinsic, unadjusted
``` 
* [Add code to incorrect `pub` restriction error](https://github.com/rust-lang/rust/pull/51463/commits/e96fdea8a38f39f99f8b9a4000a689187a457e08)
* [Add message to `rustc_on_unimplemented` attributes in core to have them set a custom message _and_ label](https://github.com/rust-lang/rust/pull/51463/commits/2cc7e5ed307aee936c20479cfdc7409d6b52a464):
```
error[E0277]: `W` does not have a constant size known at compile-time
  --> $DIR/unsized-enum2.rs:33:8
   |
LL |     VA(W),
   |        ^ `W` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `W`
   = help: consider adding a `where W: std::marker::Sized` bound
   = note: no field of an enum variant may have a dynamically sized type
```
```
error[E0277]: `Foo` cannot be sent between threads safely
  --> $DIR/E0277-2.rs:26:5
   |
LL |     is_send::<Foo>();
   |     ^^^^^^^^^^^^^^ `Foo` cannot be sent between threads safely
   |
   = help: the trait `std::marker::Send` is not implemented for `Foo`
```
```
error[E0277]: can't compare `{integer}` with `std::string::String`
  --> $DIR/binops.rs:16:7
   |
LL |     5 < String::new();
   |       ^ no implementation for `{integer} < std::string::String` and `{integer} > std::string::String`
   |
   = help: the trait `std::cmp::PartialOrd<std::string::String>` is not implemented for `{integer}`
```
```
error[E0277]: can't compare `{integer}` with `std::result::Result<{integer}, _>`
  --> $DIR/binops.rs:17:7
   |
LL |     6 == Ok(1);
   |       ^^ no implementation for `{integer} == std::result::Result<{integer}, _>`
   |
   = help: the trait `std::cmp::PartialEq<std::result::Result<{integer}, _>>` is not implemented for `{integer}`
```
```
error[E0277]: a collection of type `i32` cannot be built from an iterator over elements of type `i32`
  --> $DIR/type-check-defaults.rs:16:19
   |
LL | struct WellFormed<Z = Foo<i32, i32>>(Z);
   |                   ^ a collection of type `i32` cannot be built from `std::iter::Iterator<Item=i32>`
   |
   = help: the trait `std::iter::FromIterator<i32>` is not implemented for `i32`
note: required by `Foo`
  --> $DIR/type-check-defaults.rs:15:1
   |
LL | struct Foo<T, U: FromIterator<T>>(T, U);
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
* [Add link to book for `Sized` errors](https://github.com/rust-lang/rust/pull/51463/commits/1244dc7c283323aea1a3457a4458d590a3e160c8):
```
error[E0277]: `std::fmt::Debug + std::marker::Sync + 'static` does not have a constant size known at compile-time
  --> $DIR/const-unsized.rs:13:29
   |
LL | const CONST_0: Debug+Sync = *(&0 as &(Debug+Sync));
   |                             ^^^^^^^^^^^^^^^^^^^^^^ `std::fmt::Debug + std::marker::Sync + 'static` does not have a constant size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `std::fmt::Debug + std::marker::Sync + 'static`
   = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types--sized>
   = note: constant expressions must have a statically known size
```
* [Point to previous line for single expected token not found](https://github.com/rust-lang/rust/pull/51463/commits/48165168fb0f059d8536cd4a2276b609d4a7f721) (if the current token is in a different line)

### CODE REVIEW

**Comment by pietroalbini** [comment]
Ping from triage @nikomatsakis! This PR needs your review.

File: src/test/compile-fail/extern-types-unsized.rs [hunk 1]
```diff
@@ -30,14 +30,14 @@ fn assert_sized<T>() { }
 
 fn main() {
     assert_sized::<A>();
-    //~^ ERROR the trait bound `A: std::marker::Sized` is not satisfied
+    //~^ ERROR `A` does not have a constant size known at compile-time
```
Comment (nikomatsakis) [inline src/test/compile-fail/extern-types-unsized.rs]: something about this phrasing feels repetitive (if the size is known at compile-time, it must be constant?) or maybe just odd -- it's really that *values* of type `A` do not have a known size -- maybe just drop "constant"?

- `A` does not have a size known at compile-time

I also tried to rephrase in various ways, but I wasn't happy with them:

- the compiler cannot compute the size for values of type `A`
- values of type `A` do not have constant size
- the size for value values of type `A` cannot be known at compilation time

File: src/libsyntax/parse/parser.rs [hunk 2]
```diff
@@ -637,7 +637,26 @@ impl<'a> Parser<'a> {
                 let mut err = self.fatal(&format!("expected `{}`, found `{}`",
                                                   token_str,
                                                   this_token_str));
-                err.span_label(self.span, format!("expected `{}`", token_str));
+
+                let sp = if self.token == token::Token::Eof {
+                    // EOF, don't want to point at the following char, but rather the last token
+                    self.prev_span
```
Comment (nikomatsakis) [inline src/libsyntax/parse/parser.rs]: Is there a test reflecting this change?

**Review by nikomatsakis — CHANGES_REQUESTED** [review]

File: src/libsyntax/parse/parser.rs (hunk 2 above)
Comment (estebank) [reply src/libsyntax/parse/parser.rs]: It seems like I didn't add one. It was a drive-by fix for a diagnostic I saw while fixing the rest to unify the output with the other, more general `unexpected token` error. I'll add one after fixing the wording for `Sized`.

Comment (estebank) [reply src/libsyntax/parse/parser.rs]: Added test.

File: src/test/compile-fail/extern-types-unsized.rs (hunk 1 above)
Comment (estebank) [reply src/test/compile-fail/extern-types-unsized.rs]: Changed wording:

```
error[E0277]: the size for value values of type `str` cannot be known at compilation time
  --> $DIR/const-unsized.rs:22:26
   |
LL | static STATIC_BAR: str = *"bar";
   |                          ^^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `std::marker::Sized` is not implemented for `str`
   = note: to learn more, visit <https://doc.rust-lang.org/book/second-edition/ch19-04-advanced-types.html#dynamically-sized-types--sized>
   = note: constant expressions must have a statically known size
```

**Comment by estebank** [comment]
Ping. Anything left for me to do?

**Review by nikomatsakis — APPROVED** [review]

**Comment by kennytm** [comment]
@bors retry 

```
---- freshness::changing_bin_features_caches_targets stdout ----
running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\cargo.exe build`
running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\cit\t543\foo\target\debug\off1.exe`
running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\cargo.exe build --features foo`
running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\cit\t543\foo\target\debug\on1.exe`
running `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\release\cargo.exe build`
thread 'freshness::changing_bin_features_caches_targets' panicked at '
Expected: execs
    but: exited with exit code: 101
--- stdout
--- stderr
error: failed to remove file `C:\projects\rust\build\x86_64-pc-windows-msvc\stage2-tools\x86_64-pc-windows-msvc\cit\t543\foo\target\debug\foo.exe`
Caused by:
  Access is denied. (os error 5)
', tools\cargo\tests\testsuite\hamcrest.rs:13:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failures:
    freshness::changing_bin_features_caches_targets
test result: FAILED. 1306 passed; 1 failed; 4 ignored; 0 measured; 0 filtered out
```

File: src/libcore/marker.rs [hunk 3]
```diff
@@ -88,7 +91,12 @@ impl<T: ?Sized> !Send for *mut T { }
 /// [trait object]: ../../book/first-edition/trait-objects.html
 #[stable(feature = "rust1", since = "1.0.0")]
 #[lang = "sized"]
-#[rustc_on_unimplemented = "`{Self}` does not have a constant size known at compile-time"]
+#[rustc_on_unimplemented(
+    message="the size for value values of type `{Self}` cannot be known at compilation time",
```
Comment (lqd) [inline src/libcore/marker.rs]: is "the size for value values" a typo here ?

Comment (estebank) [reply src/libcore/marker.rs]: It is. Can you file a ticket (or even better, a quick PR)?

Comment (lqd) [reply src/libcore/marker.rs]: I've opened https://github.com/rust-lang/rust/pull/52231 to fix it

**Pull request merged**